DRILL-7926: Age Function Fix

dzamo opened the pull request:
# DRILL-7926: Age function fix
## Description

Fixed the behavior of the age function.

## Testing

1.The results of calculating the AGE function are given in accordance with PostgreSQL
2. Added 1 more test that checks the query from the issue

dzamo commented:
@paul-rogers. I think we lost Oleg on #2268.  I've closed that and opened this PR based on where he got to, plus some fixes relating to his assumptions of UTC and Unix times.  The fixes take care of the problem revealed by `select age(current_date)`, but other tests reveal another problem with java.time code used, in that intervals that span midnight are reported to be a whole day long, even when they're a few minutes.

Other changes include deletion of the harmful constants that assume a month is 30 days so that they can't do any more damage and reformating to 2-space indentation.

I'll request a review once the problem above is fixed and the tests are expanded.

oleg-zinovev commented on exec/vector/src/main/java/org/apache/drill/exec/vector/DateUtilities.java:
```diff
@@ -213,11 +216,27 @@ public static LocalDate fromDrillDate(long value) {
+   */
+  public static long toDrillTimestamp(Instant instant, ZoneId zoneId) {
+    LocalDateTime ldt = instant.atZone(zoneId).toLocalDateTime();
+    return ldt.atZone(ZoneId.of("UTC")).toEpochSecond()*1000;
```
1) I think we should use the `toInstant().toEpochMilli()`. Current implementation loses information about milliseconds.
2) `ZoneId.of ("UTC")` is a good candidate for creating a constant

dzamo replied:
Good catch, thanks.

oleg-zinovev commented:
@dzamo 

Thank you for continuing to fix the age function. Unfortunately, my work is no longer connected with Apache Drill, and it is not always possible to find personal time.

paul-rogers commented:
@dzamo, sounds like a good plan. Thanks for taking this on! Drill date handling is a bit of a pain - I appreciate your efforts to get things right. Just ping me when you're ready for a review.

dzamo commented:
Okay...

1. I converted the `AGE` function to a Freemarker template, removing duplicated code.
2. I got rid of some unnecessary time zone wrangling.  I was trying to do everything using ZonedDateTimes in the configured time zone, because of worries about edge cases like daylight savings transitions and what effect people might expect those to have on `AGE`, but I then reviewed the `TIMESTAMPDIFF` implementation, a function that tackles a similar problem, and ended up calculating only with LocalDateTimes, consistent with `TIMESTAMPDIFF`.
3. The basic solution is Oleg's final Java Time API one, which is I think is very clean, and it works.  The issue I thought it had over midnights was a red herring due to the string literals I was testing with get truncated to dates for lacking a `timestamp` prefix.  So really the only thing wrong with where Oleg got to was the time zone handling of the query start time used in the unary form of `AGE`.
4. I've added tests, mainly for the unary form.
5. The Java Time techniques used here present an opportunity to considerably simplify the Java Time part of `TIMESTAMPDIFF`, so I did that.  It's in a separate commit and doesn't have to be part of this PR if that would be cheating on paperwork.

paul-rogers commented on exec/java-exec/src/main/codegen/templates/DateIntervalFunctionTemplates/AgeFunction.java:
```diff
@@ -0,0 +1,106 @@
+
+    public void eval() {
+      java.time.LocalDateTime from = java.time.Instant.ofEpochMilli(right.value).atZone(java.time.ZoneOffset.UTC).toLocalDateTime();
+      java.time.LocalDateTime to = java.time.Instant.ofEpochMilli(left.value).atZone(java.time.ZoneOffset.UTC).toLocalDateTime();
```
Boy, this is a tricky area! I'm not sure that this does what we want. The incoming values are assumed to be in local time. We convert them, in place, to a UTC time, with time zone. We then convert that to a local time, stating that the values are UTC. If I have a time that is noon, it gets converted to noon, UTC. That is then converted to my local time, UTC-8, and will thus be 8 PM. That's not what we want. We want noon local time.

To see this, just run the above code in a unit test on a machine with a time zone other than UTC. Compare the timestamp value before and after the conversion. I suspect they will be different.

A quick check of the Javadoc did not reveal an obvious method to use to go from a local timestamp to a `LocalDateTime`. This stuff always requires lots of fiddling (at least for me) to find the right combination.

dzamo replied:
@paul-rogers

Recall my comment on the previous PR where I observe that the expressions above _do_ yield the correct LocalDateTime when you start from milliseconds since the "local epoch".  If they didn't work there would be widespread breakage in Drill, e.g. `select cast(now() as varchar)` would give the wrong answer except in Drillbits set to UTC because it relies on identical code (c.f. CastDateVarChar.java).  Yet said query correctly displays the local time on my UTC+2 machine.

Here's my corrected sequence for your UTC-8 example.  The incoming value is known to be milliseconds since the local epoch i.e. midnight 1970-01-01 in UTC-8.  We deliberately misconstrue it as Unix time using `ofEpochMilli`, producing an absolute time with an error of -8 hours.  We express that in the UTC time zone using `atZone`, then throw away the zone component using `toLocalDateTime`, then finally declare that result to be in UTC-8, picking up a compensating error of +8 hours and the correct local time.

I'm not advocating the above!  I did not invent it, I merely encountered it (as did the unfortunate Oleg).  I wish I could unsee it and purge it from my consciousness but 

- it does give the correct answer,
- it is used fairly widely in the code base making it kind of a _de facto_ standard,
- included in the prior point are a Joda equivalent, and instances of the same procedure in reverse (starting with a local date object and ending in ms since "local epoch"),
- the unfortunate ms since the "local epoch" representation is probably always going to require _something_ nasty and confusing.

I think replacing this date code pattern in the code base, if we do, should be a fell swoop achieved in another PR devoted entirely to date code refactoring (and which also replaces usage of Joda).  In the short term I propose that to fix the badly broken `AGE` function we should resign ourselves to using date code consistent with what's already there.

paul-rogers commented on exec/java-exec/src/main/codegen/templates/DateIntervalFunctionTemplates/AgeFunction.java:
```diff
@@ -0,0 +1,106 @@
+  public static class Age${fromUnit}ToMidnight implements DrillSimpleFunc {
+
+    @Param ${fromUnit}Holder right;
+    @Workspace java.time.LocalDateTime to;
```
Suggestion: explain the logic here. Not sure anyone not in on our conversation will understand what we're trying to do.

paul-rogers commented on exec/java-exec/src/main/codegen/templates/DateIntervalFunctionTemplates/AgeFunction.java:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    public void eval() {
+      java.time.LocalDateTime from = java.time.Instant.ofEpochMilli(right.value).atZone(java.time.ZoneOffset.UTC).toLocalDateTime();
```
Interestingly, here we're using an actual UTC-based Unix timestamp, so the conversion is correct. A comment to explain this will help future developers.

dzamo replied:
The _query start time_ higher up is indeed a real Unix timestamp, and converted to a LocalDateTime differently.  In the excerpted lines we've got another instance of a Drill timestamp in ms since the "local epoch" and the same date code pattern described in my earlier comment...

paul-rogers commented on exec/java-exec/src/main/codegen/templates/DateIntervalFunctionTemplates/TimestampDiffFunction.java:
```diff
@@ -72,25 +72,11 @@ public void eval() {
-      int months = between.getMonths() + between.getYears() * org.apache.drill.exec.vector.DateUtilities.yearsToMonths;
+
+      java.time.LocalDateTime from = java.time.Instant.ofEpochMilli(left.value).atZone(java.time.ZoneOffset.UTC).toLocalDateTime();
+      java.time.LocalDateTime to = java.time.Instant.ofEpochMilli(right.value).atZone(java.time.ZoneOffset.UTC).toLocalDateTime();
```
Again.

dzamo replied:
See my first comment.

paul-rogers commented on exec/java-exec/src/main/codegen/templates/DateIntervalFunctionTemplates/AgeFunction.java:
```diff
@@ -0,0 +1,106 @@
+    public void setup() {
+      java.time.ZoneId zoneId = contextInfo.getRootFragmentTimeZone();
+      java.time.ZonedDateTime zdtStart = contextInfo.getQueryStartInstant().atZone(zoneId);
+      to = zdtStart.truncatedTo(java.time.temporal.ChronoUnit.DAYS).toLocalDateTime();
```
Interestingly, here we're using an actual UTC-based Unix timestamp, so the conversion used earlier is the correct one. But, here, are we doing the right thing? We want a local date/time with the end date, as converted from a UTC timestamp. Instead, it seems we are converting from local time to UTC.

Are we trying to do the age calcs in UTC? We should do them in local time so that a baby born at 11:00PM local time yesterday (the data value) will have an age of 2 hours for a query run at 1AM today.

(This is where that definition comment would be handy: the common convention is that the baby is one day old at midnight, though in our terms, at midnight the baby is one hour old. Thus, in our convention, the person won't turn 21, say, until 11PM on their 21th birthday, though a bartender will happily serve them any time on that day.)

A comment to explain this will help future developers.

dzamo replied:
Yes here is indeed a real Unix timestamp.  Actually, other than that, UTC does _not_ feature anywhere here.  Instead we convert this query start time to the the local time zone (`getRootFragmentTimeZone`) get to a LocalDateTime.  The age calcs all happen with LocalDateTimes, not UTC or any time zone, as they do in e.g. `TIMESTAMPDIFF` (including before I refactored it).  If we want to start using zoned date time objects for these calcs I think we should open a discussion about consistently applying that across all functions (so at least also `TIMESTAMPDIFF`) in a separate PR.

At 1am local time, a baby born yesterday at 11pm is reported as being one hour old by the unary form of `AGE` (substract from midnight) and two hours old by the binary invocation `select AGE(current_timestamp, <baby's birth time>`.

I've added comments explaining that AGE does not count calendar boundary crossings, like midnight, as the barman does.

paul-rogers commented on exec/java-exec/src/main/codegen/templates/DateIntervalFunctionTemplates/AgeFunction.java:
```diff
@@ -0,0 +1,106 @@
+<#list dateIntervalFunc.dates as fromUnit>
+<#list dateIntervalFunc.dates as toUnit>
+
+  @FunctionTemplate(name = "age",
```
Would be worth adding a comment to state that this is for the two-argument AGE function, and that we are following Postgres. Might even want to include a link to the Postgres definition. (Which seems to be that `AGE` is the same as `to - from`.)

paul-rogers reviewed (commented):
Thanks for tackling this monster. I suspect that either some of the conversions need a bit more attention or that I'm confused. It is an exercise for the reader to determine which.

paul-rogers reviewed (approved):
@dzamo thanks for the explanations. No matter how much I fight with Drill's odd date-time code, it is a new battle each time. Seems like you've got the logic sorted out. Thanks much for your efforts.
LGTM +1

dzamo commented:
Thanks @paul-rogers for the (considerable) work done reviewing the attempts made on this one.
